### PR TITLE
Templatise `rorN` and `rolN` with `N` in {8, 16, 32, 64}

### PR DESCRIPTION
**Pull request by tomlankhorst**
Reduce code duplication. 
Allows arbitrary unsigned integral type rotations. 

Useful for the upcoming templatized `ror-xor-ror` PRNG. 

E.g.
```cpp
template<typename _Type>
function magic(_Type x) {
  return rol<_Type>(x,5);
}
```

### CODE REVIEW

**Comment by Gymnasiast** [comment]
I'm afraid I have to agree with Aaron here. Especially as the current patterns are recognised by most compilers as being `rol`/`ror` instructions.

**Comment by tomlankhorst** [comment]
> In fact, it's now over three times longer than it was...
It is, but that's mainly because I defined function aliases for the existing rolN/rorN. 

> Especially as the current patterns are recognised by most compilers as being rol/ror instructions.
That's not changed very much. The two version produce two times:
```
        mov     eax, edi
        mov     ecx, edi
        ror     eax, cl
```
`GCC 7.1  -O1 -std=c++17`

**Comment by AaronVanGeffen** [comment]
> It is, but that's mainly because I defined function aliases for the existing rolN/rorN.

Fair point. Your code also allows us to drop the integer and infer the amount of bits to rotate based on the type in question, of course. (That is, assuming we don't apply `rot8` to a `uint16_t`...)

**Comment by IntelOrca** [comment]
I like this change.

**Comment by Gymnasiast** [comment]
Well, in that case, I have no objections.

**Comment by IntelOrca** [comment]
@AaronVanGeffen 

> (That is, assuming we don't apply rot8 to a uint16_t...)

Well you would not be able to if you remove the ror8 altogether as `ror(some uint16)` would give you ror16 automatically.

**Comment by tomlankhorst** [comment]
What's also cool is that the template type can be deducted from the first parameter:
```
uint32_t a = 0x100;
auto b = rol(a,8); // 0x10000
```

**Comment by Gymnasiast** [comment]
From what I gather, this can be merged? @IntelOrca

**Comment by IntelOrca** [comment]
Depending on how big of a change that would be to do, it could be a follow up PR.

**Comment by tomlankhorst** [comment]
Like, openrct2/util/Rotate.h?

**Comment by Gymnasiast** [comment]
.h_pp_ , but yeah.

**Comment by IntelOrca** [comment]
or if we wanted to generalise more, `Numerics.hpp`

**Comment by IntelOrca** [comment]
Would you be able to move the hpp to core? util is sort of obsolete now.

File: src/openrct2/common.h [hunk 1]
```diff
@@ -41,14 +43,14 @@ using utf16string = utf16*;
...
 
-#define rol8(x, shift) (((uint8_t)(x) << (shift)) | ((uint8_t)(x) >> (8 - (shift))))
-#define ror8(x, shift) (((uint8_t)(x) >> (shift)) | ((uint8_t)(x) << (8 - (shift))))
-#define rol16(x, shift) (((uint16_t)(x) << (shift)) | ((uint16_t)(x) >> (16 - (shift))))
-#define ror16(x, shift) (((uint16_t)(x) >> (shift)) | ((uint16_t)(x) << (16 - (shift))))
-#define rol32(x, shift) (((uint32_t)(x) << (shift)) | ((uint32_t)(x) >> (32 - (shift))))
-#define ror32(x, shift) (((uint32_t)(x) >> (shift)) | ((uint32_t)(x) << (32 - (shift))))
-#define rol64(x, shift) (((uint64_t)(x) << (shift)) | ((uint64_t)(x) >> (64 - (shift))))
-#define ror64(x, shift) (((uint64_t)(x) >> (shift)) | ((uint64_t)(x) << (64 - (shift))))
+const constexpr auto rol8 = rol<uint8_t>;
```
Comment (Gymnasiast) [inline src/openrct2/common.h]: I think (but I'd like him to confirm it) that @IntelOrca also meant that these consts should be moved.

Comment (janisozaur) [reply src/openrct2/common.h]: :+1:

Comment (IntelOrca) [reply src/openrct2/common.h]: It doesn't really matter since they are going to be deleted eventually.

File: src/openrct2/core/Numerics.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#pragma once
```
Comment (janisozaur) [inline src/openrct2/core/Numerics.hpp]: Missing copyright header. When copying, remember it's 2019

Comment (tomlankhorst) [reply src/openrct2/core/Numerics.hpp]: Sure. Will add it to my IDE's template while I'm at it. 
Are we going to update the year numbering every now and then? 
I feel like that's horrible for version control and CI.

**Pull request merged**